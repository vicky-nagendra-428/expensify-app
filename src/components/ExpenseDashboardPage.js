import React from 'react';
import ExpensesList  from './ExpenseList';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import numeral from 'numeral';
import getFilteredExpenses from '../selectors/expenses';
import getExpensesSum from '../selectors/expensesSum';

class ExpenseDashboardPage extends React.Component {
    state = {
        focusedInput: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }
    onFocusChange = (focusedInput) => {
        this.setState(() => {
            return {
                focusedInput
            }
        })
    }
    getExpensesSum = (expenses, filters) => {
        
        const filteredExpenses = getFilteredExpenses(expenses, filters);
        const filteredExpensesSum = getExpensesSum(filteredExpenses);

        return <p>You are viewing {filteredExpenses.length} expenses, sum to {numeral(filteredExpensesSum / 100).format('$0,0.00')}</p>
        
    }
    render() {
        return (
            <div>
                <p>This is expense dashboard page</p>
                {this.getExpensesSum(this.props.expenses, this.props.filters)}
                <input type="text" value={this.props.filters.text}
                    onChange={(e) => {
                        this.props.dispatch(setTextFilter(e.target.value));
                        // dispatch is sent by the connect,
                    }}
                />
                sortBy
                <select
                    value={this.props.filters.sortBy}
                    onChange={(e) => {
                        e.target.value === 'amount' ? this.props.dispatch(sortByAmount()) : this.props.dispatch(sortByDate())
                    }}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="ids1"
                    endDate={this.props.filters.endDate}
                    endDateId="ide1"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
                <ExpensesList />
            </div>
        );
    } 
}

const connectReactRedux = (state) => {
    return {
        expenses: state.expenses,
        filters: state.filters
    }
};

export default connect(connectReactRedux)(ExpenseDashboardPage);