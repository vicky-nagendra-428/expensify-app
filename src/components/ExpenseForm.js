import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        }
    }

    // state = {
    //     description: this.props.expense ? this.props.expense.description : '',
    //     note: this.props.expense ? this.props.expense.note : '',
    //     amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    //     createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    //     calendarFocused: false,
    //     error: undefined
    // }
    
    onDescriptionChnage = (e) => {
        const description = e.target.value
        this.setState(() => {
            return {
                description
            }
        })
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => {
            return {
                note
            }
        })
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => {
                return {
                    amount
                }
            })
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(() => {
                return {
                    error: 'Please fill all mandatory fields'
                }
            })
        } else {
            this.setState(() => {
                return {
                    error: ''
                }
            })
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                <h1>Expense Form</h1>
                <form
                    onSubmit={this.onSubmit}
                >
                    <input 
                        type="text"
                        placeholder="description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChnage}
                    />
                    <input 
                        type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <textarea 
                        placeholder="more notes"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id="id1"
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <button
                        type="submit"
                    >Add Expense</button>
                </form>
                {this.state.error && <p>{this.state.error}</p>}
            </div>
        )
    }
};