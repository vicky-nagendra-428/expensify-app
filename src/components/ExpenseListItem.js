import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({id, description, amount, createdAt}) => (
    <div>
        <p>
            <Link to={`/edit/${id}`}>{description}</Link>
        </p>
        <p>{amount}</p>
        <p>{createdAt}</p>
       
    </div>
);

export default ExpenseListItem;