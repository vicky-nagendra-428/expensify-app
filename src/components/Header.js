import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = (props) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" exact={true} activeClassName="is-active">Home page</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <NavLink to="/edit/:id" activeClassName="is-active">Edit Expense</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <button
            onClick={props.startLogout}
        >Logout</button>
    </header>
);

const mapDispatchToProps = (dispatch) => {
    return {
        startLogout: () => dispatch(startLogout())
    }
}

export default connect(undefined, mapDispatchToProps)(Header);