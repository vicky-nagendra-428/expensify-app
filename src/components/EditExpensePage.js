import React from 'react';
import  { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onUpdate = (expense) => {
        this.props.onSubmit(this.props.expense.id, expense)
        this.props.history.push("/")
    }
    onRemove = () => {
        this.props.onClick({id: this.props.expense.id})
        this.props.history.push("/")
    } 
    render() {
        return (
            <div>
                <h3>Edit expense</h3>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onUpdate}
                />
                <button
                    onClick={this.onRemove}
                >remove</button>
            </div>
        );
    }
} 

const connectReduxStore = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}

const connectDispatchToProps = (dispatch, props) => {
    return {
        onSubmit: (id, expense) => {
            dispatch(startEditExpense(id, expense))
        },
        onClick: (id) => {
            dispatch(startRemoveExpense(id))
        }
    }
}

export default connect(connectReduxStore, connectDispatchToProps)(EditExpensePage);