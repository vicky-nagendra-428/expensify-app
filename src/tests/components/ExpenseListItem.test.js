import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import moment from 'moment';

test('expense list item', () => {
    const item = {
        id: '1',
        description: 'description',
        amount: 120,
        createdAt: 100
    }
    const wrapper = shallow(<ExpenseListItem {...item}/>);
    expect(wrapper).toMatchSnapshot();
})