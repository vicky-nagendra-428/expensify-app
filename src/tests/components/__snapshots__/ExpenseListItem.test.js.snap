// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expense list item 1`] = `
<div>
  <p>
    <Link
      to="/edit/1"
    >
      description
    </Link>
  </p>
  <p>
    $1.20
  </p>
  <p>
    January 1st 1970
  </p>
</div>
`;
